### PR TITLE
[DQL2-6] download archive to a temporary file so we can do type sniffing

### PR DESCRIPTION
**Pull request by ThrawnCA**
- This uses the external URL without an API key, so it will not work on private datasets, but is otherwise fairly reliable.

Most of the download code was copied from `ckanext-xloader`.

### CODE REVIEW

File: ckanext/qa/tasks.py [hunk 1]
```diff
@@ -356,35 +369,147 @@ def score_by_sniffing_data(archival, resource, score_reasons):
...
-            score = lib.resource_format_scores().get(sniffed_format['format']) \
-                if sniffed_format else None
-            if sniffed_format:
-                score_reasons.append(_('Content of file appeared to be format "%s" which receives openness score: %s.')
-                                     % (sniffed_format['format'], score))
-                return score, sniffed_format['format']
-            else:
-                score_reasons.append(_('The format of the file was not recognized from its contents.'))
+        log.debug("File not found on disk for resource %s", resource)
+        if resource.url_type == 'upload':
```
Comment (duttonw) [inline ckanext/qa/tasks.py]: if type url, then user archive url if it exists. this is missing.

Comment (duttonw) [reply ckanext/qa/tasks.py]: try archival.cache_url for download if url

File: ckanext/qa/tasks.py [hunk 2]
```diff
@@ -356,35 +369,147 @@ def score_by_sniffing_data(archival, resource, score_reasons):
...
+            # i.e. 'Download failure' or 'System error during archival'
+            score_reasons.append(_('A system error occurred during downloading this file') + '. '
+                                 + _('Reason') + ': %s. ' % archival.reason + _('Using other methods to determine file openness.'))
+            return (None, None)
+        else:
+            score_reasons.append(_('This file had not been downloaded at the time of scoring it.'))
+            return (None, None)
+
+
+def _download_url(url):
```
Comment (duttonw) [inline ckanext/qa/tasks.py:417]: could the download url code be in a helper class?

**Review by duttonw — CHANGES_REQUESTED** [review]
good that it handles upload, do add the change for archive.cache_url for url_type = url so we handle both types.

**Review by duttonw — APPROVED** [review]

File: ckanext/qa/tasks.py (hunk 2 above)
Comment (ThrawnCA) [reply ckanext/qa/tasks.py:417]: It could, if there was a good case for sharing it with other code. At this point that doesn't really seem practical.

**Pull request merged**